### PR TITLE
Basic data type CRU methods

### PR DESCRIPTION
**Pull request by benwerner01**
## 🌟 What is the purpose of this PR?

This PR introduces the basic create, read and update (CRU) datastore methods for data types.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1201095311341924/1202510174412968/f) _(internal)_

## 🔍 What does this change?

- implements `create_data_type`, `get_data_type` and `update_data_type` methods

## 📜 Does this require a change to the docs?

Not that I'm aware of

## 🐾 Next steps

- Figure out how primitive data types are created in the datastore

## 🛡 What tests cover this?

- unit tests

## ❓ How to test this?

1. Setup the local db
2. Run the `PostgresDatabase` tests

### CODE REVIEW

File: packages/graph/hash_graph/lib/graph/src/datastore/mod.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/datastore/mod.rs:171]:
```suggestion
    /// Update the definition of an existing [`DataType`].
```

File: packages/graph/hash_graph/lib/graph/src/datastore/mod.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/datastore/mod.rs:175]: Not requesting a change but I wonder if we should consider `update` operations that create if it does not exist.

I find it really annoying to have to do:
Read to check and then
Create or Update

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:36]: @TimDiekmann how do we feel about documenting functions like this? I think we generally want all pub functions documented, but I don't think we should _never_ document private functions either.

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:45]: **_Tiny_** nit: I think I'd do these two attachments in the other order

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:85]: Happy to leave this as is for this PR because it'll be easier to swap in another implementation, but one of the benefits of moving this into a method (outside of this) is that we can reuse it for the other "create" methods, and importantly we want to think about sortable version_ids if we can. As such, the `new` method would need to know about the older one (if it's not time based) so it can increment.

Lots to unpack, I don't think we should do anything in this PR about it.

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:140]: Just flagging that this sort of thing is why having clearer methods, e.g. `Identifier::new` or `Identifier::from` would be more clear here, because at first glance this looks very similar to the one used in `create_data_type` and if skimming quickly it might be easy to assume it's the same and isn't reusing the base_id.

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:147]: I don't have a good suggestion for this so perhaps it's fine as is (for now). But it feels kind of wasteful to me to be returning these structs and to immediately ignore most of it, rather than only allocating and returning the necessary piece.

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:107]: I'm not sure how I feel about writing our SQL inline in strings in functions. In the long run it'd be good to see if we can get these version controlled in actual `sql` files or something so we can benefit from more SQL tooling like linting, intellisense, etc.

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:302]:
```suggestion
        let data_type = db.get_data_type(id).await?;
```

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:322]:
```suggestion
            .await?;
```

File: packages/graph/hash_graph/lib/graph/src/types/mod.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/types/mod.rs:51]:
```suggestion
        // TODO: change this to not just using the debug impl.
```

Why'd we remove the version you guys had when we were on a call? I know I raised some issues with it but like I said, I thought it was fine for now :)

**Review by Alfred-Mountfield — CHANGES_REQUESTED** [review]
Looking good, a couple of small things.

I've also added a lot of comments about bits I think we should think about as we keep developing, they do _not_ need to be addressed as part of this PR. (And in fact I think they _shouldn't_ be. Although if we do think some of them are valuable then perhaps we encapsulate them as `TODO`s)

File: packages/graph/hash_graph/lib/graph/src/datastore/mod.rs
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/datastore/mod.rs:155]:
```suggestion
    /// - [`DatastoreError`], if the account referred to by `created_by` does not exist.
```

File: packages/graph/hash_graph/lib/graph/src/datastore/mod.rs
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/datastore/mod.rs:166]:
```suggestion
    /// - [`DatastoreError`], if the [`DataType`] doesn't exist.
```

File: packages/graph/hash_graph/lib/graph/src/datastore/mod.rs
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/datastore/mod.rs:175]:
```suggestion
    /// - [`DatastoreError`], if the [`DataType`] doesn't exist.
```

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:24]:
```suggestion
    /// - [sqcx `Error`], if creating a [`PgPool`] connection returns an error.
    ///
    /// [sqcx `Error`]: sqlx::Error
```

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:271]: I think we should wait merging this PR until we can run the tests in CI (https://github.com/hashintel/hash/pull/733).
```suggestion
    #[tokio::test]
    #[cfg_attr(miri, ignore = "miri can't run in async context")]
```

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:288]:
```suggestion
    #[tokio::test]
    #[cfg_attr(miri, ignore = "miri can't run in async context")]
```

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:310]:
```suggestion
    #[tokio::test]
    #[cfg_attr(miri, ignore = "miri can't run in async context")]
```

File: packages/graph/hash_graph/lib/graph/src/types/mod.rs
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/types/mod.rs:9]: ***Tiny*** nit: putting the `transparent` attributes next to each other looks better. Also I'd prefer the derived traits to list the std-ones first, then third party ones. Personally, I always use this order: `Debug`, `Copy`, `Clone`, comparisons (`PartialEq`, `Eq`, `PartialOrd`, `Ord`), `Hash`, but that's a matter of taste. Feel free to ignore these suggestions.
```suggestion
#[repr(transparent)]
#[sqlx(transparent)]
#[derive(Debug, Copy, Clone, PartialEq, Eq, sqlx::Type)]
```

File: packages/graph/hash_graph/lib/graph/src/types/mod.rs
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/types/mod.rs:21]:
```suggestion
#[repr(transparent)]
#[sqlx(transparent)]
#[derive(Debug, Copy, Clone, PartialEq, Eq, sqlx::Type)]
```

File: packages/graph/hash_graph/lib/graph/src/types/mod.rs
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/types/mod.rs:33]:
```suggestion
#[repr(transparent)]
#[sqlx(transparent)]
#[derive(Debug, Copy, Clone, PartialEq, Eq, sqlx::Type)]
```

File: packages/graph/hash_graph/lib/graph/src/datastore/mod.rs
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/datastore/mod.rs:175]: I wonder if we can leverage Postgres' [`ON CONFLICT`](https://www.postgresql.org/docs/current/sql-insert.html#SQL-ON-CONFLICT) upserts to accomplish this behavior in SQL

File: packages/graph/hash_graph/lib/graph/src/types/mod.rs
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/types/mod.rs:58]:
```suggestion
#[repr(transparent)]
#[sqlx(transparent)]
#[derive(Debug, Clone, PartialEq, Eq, sqlx::Type)]
```

File: packages/graph/hash_graph/lib/graph/src/types/mod.rs
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/types/mod.rs:72]: Maybe rename to `value`?

File: packages/graph/hash_graph/lib/graph/src/types/mod.rs
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/types/mod.rs:73]: I feel like these fields shouldn't be public

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:282]: I think these `expect`s were added because the `PostgresDatabase::new` returns a `sqlx::Error` in the error case, which is not `DatastoreError` the rest returns. Should `PostgresDatabase::new` be changed, or do we just expect on that instead?

**Review by TimDiekmann — CHANGES_REQUESTED** [review]
Overall PR looks good. 🎉 

mainly convention suggestions and a few ideas. We should wait for #733 to test the functionality in CI as well before merging.

File: packages/graph/hash_graph/lib/graph/src/types/mod.rs
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/types/mod.rs:51]: Tim pointed out the output looked almost identical to the default debug print, so we opted for that instead (just after you left the call :sweat_smile: )

File: packages/graph/hash_graph/lib/graph/src/datastore/mod.rs
Comment (benwerner01) [reply packages/graph/hash_graph/lib/graph/src/datastore/mod.rs:175]: My instinct is that it would be better to have a separate `updateOrCreate` method that encapsulates this logic and the calls to `update` and `create`, as it could be confusing for this to be the default behaviour.

A caller of the method may be unaware they are passing the wrong id, think they're correctly updating an existing type, but instead they've ended up creating a new entity in the graph.

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (benwerner01) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs]: Addressed in https://github.com/hashintel/hash/pull/734/commits/811e86c8c3ecec352ec9f303d7773b124697fc9d, can you help me understand why this ordering is important?

File: packages/graph/hash_graph/lib/graph/src/datastore/mod.rs
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/datastore/mod.rs:175]: That's also a valid point Ben. The difference in signature with the IDs might make this a little questionable. I think it'll be a lot harder to accidentally push a mistyped URI though.

Lets discuss more later :) As I said, definitely no need to address in this PR

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs]: When you print the report, it prints out the entire stack, and it respects the order in which you built it. As such you kind of want to build the context incrementally

- Some Error Happened
- 318903180-39131-1313801
- Could not insert id

Seems a little weird because you just have a random ID, as opposed to

- Some Error Happened
- Could not insert id
- 318903180-39131-1313801

It's not super important, and the printing might even be reversed. I just think it's good practice to say the message first, and get more specific after.

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:24]: sqlx* Not sqcx.

I also think we should probably return `DatastoreError` here, or something like `PostgresDbError`

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:282]: There's a comment thread on the PostgresDatabase module where I suggested we should probably change the error types. However, in these tests we can just call `change_context(DatastoreError)`

File: packages/graph/hash_graph/lib/graph/src/types/mod.rs
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/types/mod.rs:9]: `sqlx::Type` has to come before `sqlx(transparent)` I believe.

File: packages/graph/hash_graph/lib/graph/src/datastore/mod.rs
Comment (benwerner01) [reply packages/graph/hash_graph/lib/graph/src/datastore/mod.rs]: Addressed in https://github.com/hashintel/hash/pull/734/commits/4b14fd9e283d0d74de3caaf3a34e165e7c318c09

File: packages/graph/hash_graph/lib/graph/src/datastore/mod.rs
Comment (benwerner01) [reply packages/graph/hash_graph/lib/graph/src/datastore/mod.rs]: Addressed in https://github.com/hashintel/hash/pull/734/commits/4b14fd9e283d0d74de3caaf3a34e165e7c318c09

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (benwerner01) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs]: Changed this to propagate `DatastoreError` instead

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (benwerner01) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs]: Addressed in https://github.com/hashintel/hash/pull/734/commits/cb7380d5a924ca58377712f530db8280e82c3c5f (or a prior commit can't recall)

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (benwerner01) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs]: Addressed in https://github.com/hashintel/hash/pull/734/commits/f50585663a877a9865d5ce5ecb7a264ecbca1eae

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (benwerner01) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs]: Addressed this in https://github.com/hashintel/hash/pull/734/commits/50cce2ec5405d28a0e193555114bd882f18e0094

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:107]: Just for completeness, sqlx does have a macro that imports from sql files, [`sqlx::query_file_as!`](https://docs.rs/sqlx/latest/sqlx/macro.query_file_as.html). Those do compile-time checks on the query AFAIK.

File: packages/graph/hash_graph/lib/graph/src/types/mod.rs
Comment (benwerner01) [reply packages/graph/hash_graph/lib/graph/src/types/mod.rs]: Addressed in https://github.com/hashintel/hash/pull/734/commits/5b3ca35ed8893e86dbf36467a512d8f8e1af1c9d

File: packages/graph/hash_graph/lib/graph/src/types/mod.rs
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/types/mod.rs:9]: Hmm, that makes sense but annoys me 😄

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:36]: I wasn't sure how much value docs would add on these but seeing the changes I actually think this is much better. Thanks :) (although explicitly talking about `version_id` and `base_id` does mean we need to make sure to update these docs when we change how versioning is done in the table schemas

**Review by Alfred-Mountfield — APPROVED** [review]
Have a few other thoughts on improving but let's not over-optimise. This is a great foundation, changes from my side look good, will let @TimDiekmann affirm his comments are addressed too.

File: packages/graph/hash_graph/lib/graph/src/types/mod.rs
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/types/mod.rs:8]: `#[repr(transparent)]` isn't **needed**, no. But because we have `sqlx::transparent` it seems like we should also have it? This is basically just a type alias (with type-safety) for the UUID that lives inside the DB.

Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/types/mod.rs:8]: I see that type safety is the main reason we have this type. But semantically I don't see a reason, why an account id should be copied but borrowed instead. 

I like to **not** block this PR on that.

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/types/mod.rs:8]: Gotcha, let's follow-up on a call

**Comment by TimDiekmann** [comment]
The only things left for me are mainly about general code philosophy or personal taste, so I don't see a reason blocking this PR. 
I agree that this is a great foundation 🎉

**Review by TimDiekmann — APPROVED** [review]

**Pull request merged**
